pkg/termination/termination.go

```diff
@@ -0,0 +1,155 @@
+package termination
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	gcpTerminationEndpointURL = "http://169.254.169.254/computeMetadata/v1/instance/preempted"
+)
+
+// Handler represents a handler that will run to check the termination
+// notice endpoint and delete Machine's if the instance termination notice is fulfilled.
+type Handler interface {
+	Run(stop <-chan struct{}) error
+}
+
+// NewHandler constructs a new Handler
+func NewHandler(logger logr.Logger, cfg *rest.Config, pollInterval time.Duration, namespace, nodeName string) (Handler, error) {
+	c, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating client: %v", err)
+	}
+
+	pollURL, err := url.Parse(gcpTerminationEndpointURL)
+	if err != nil {
+		// This should never happen
+		panic(err)
+	}
+
+	logger = logger.WithValues("node", nodeName, "namespace", namespace)
+
+	return &handler{
+		client:       c,
+		pollURL:      pollURL,
+		pollInterval: pollInterval,
+		nodeName:     nodeName,
+		namespace:    namespace,
+		log:          logger,
+	}, nil
+}
+
+// handler implements the logic to check the termination endpoint and delete the
+// machine associated with the node
+type handler struct {
+	client       client.Client
+	pollURL      *url.URL
+	pollInterval time.Duration
+	nodeName     string
+	namespace    string
+	log          logr.Logger
+}
+
+// Run starts the handler and runs the termination logic
+func (h *handler) Run(stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	errs := make(chan error, 1)
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		errs <- h.run(ctx, wg)
+	}()
```

Comment:
I realise I wrote this for the AWS one, but looking back, I think this would be better to defer the wg.Done() here and not pass it into run, like this:
(When I was reviewing, I was like, "woah, where's the `wg.Done()`")

```suggestion
	go func() {
		defer wg.Done()
		errs <- h.run(ctx)
	}()
```